modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/syncfs/SyncFsUtils.java

```diff
@@ -0,0 +1,146 @@
+     */
+    public static final String IGNITE_CHECKPOINT_SYNC_FS_ENABLED = "IGNITE_CHECKPOINT_SYNC_FS_ENABLED";
+
+    /**
+     * Name of the system property that controls the threshold for the number of files that will trigger the
+     * {@code syncfs} instead of multiple {@code fsync} calls for each individual file in the checkpoint.
+     */
+    public static final String IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD = "IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD";
+
+    /**
+     * Value of the {@link #IGNITE_CHECKPOINT_SYNC_FS_ENABLED} system property. {@code true} by default.
+     */
+    public static final boolean SYNC_FS_ENABLED = getBoolean(IGNITE_CHECKPOINT_SYNC_FS_ENABLED, true);
+
+    /**
+     * Value of the {@link #IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD} system property. {@code 1024} by default.
+     */
+    public static final int SYNC_FS_THRESHOLD = getInteger(IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD, 1024);
+
+    /**
+     * Executes {@code syncfs} on all filesystems that contain files from the given set of page stores.
+     * According to {@code man}:
+     * <p/><i>syncfs() is like sync(), but synchronizes just the filesystem containing file referred to by the open file
+     * descriptor fd.<br/>
+     * ... sync() or syncfs() provide the same guarantees as fsync() called on every file in the system or
+     * filesystem respectively.</i>
+     *
+     * @param updStores Internal map from the checkpointer that contains all updated page stores.
+     * @throws IgniteCheckedException If operation failed.
+     */
+    public static void syncfs(
+        ConcurrentMap<PageStore, CheckpointPageStoreInfo> updStores
+    ) throws IgniteCheckedException {
+        IntHashMap<IgniteBiTuple<Path, FileStore>> groupIdToFileStore = new IntHashMap<>();
+
+        // First pass is to split all file stores by the groupId. All partitions within a group are always in the same
+        // directory.
+        for (Map.Entry<PageStore, CheckpointPageStoreInfo> entry : updStores.entrySet()) {
+            CheckpointPageStoreInfo info = entry.getValue();
+
+            if (groupIdToFileStore.containsKey(info.groupId))
+                continue;
+
+            Path path = ((FilePageStore) entry.getKey()).getPath();
+
+            try {
+                groupIdToFileStore.put(
+                    info.groupId,
+                    new IgniteBiTuple<>(path, Files.getFileStore(path))
+                );
+            }
+            catch (IOException e) {
+                throw new StorageException(e);
+            }
+        }
+
+        // Maps file store name to a path from that file store. Second pass is to fill this map.
+        Map<String, Path> fileSystems = new HashMap<>();
+
+        groupIdToFileStore.forEach((key, val) -> fileSystems.putIfAbsent(val.getValue().name(), val.getKey()));
+
+        // Resulting collection should not be big. For most installations it will only consist of a single element.
+        for (Path path : fileSystems.values()) {
```

Comment:
I suggest to separate this method in two. First half (extracted to a different method) would form the collection that you obtain via `fileSystems.values()` from `updStores` (for example, `fileSystemsToSync()`); second half (again, a new method) would sync the filesystems in that collection (for example, `syncFileSystems()`). Current method is very long..

Reply:
Ok